### PR TITLE
XHR: support typed arrays for request payloads

### PR DESCRIPTION
**Pull request by philikon**
Support `xhr.send(data)` for typed arrays.

**Test plan:** run UIExplorer example on iOS and Android.

### CODE REVIEW

**Comment by philikon** [comment]
NB: currently inclues #11901, will rebase once that's landed.

File: Libraries/WebSocket/WebSocket.js [hunk 1]
```diff
@@ -108,7 +97,7 @@ class WebSocket extends EventTarget(...WEBSOCKET_EVENTS) {
     this._close(code, reason);
   }
 
-  send(data: string | ArrayBuffer | ArrayBufferView): void {
```
Comment (philikon) [inline Libraries/WebSocket/WebSocket.js]: I'll fix those with the next rebase. (Need #11904 to land first.)

File: Libraries/WebSocket/WebSocket.js [hunk 2]
```diff
@@ -118,15 +107,8 @@ class WebSocket extends EventTarget(...WEBSOCKET_EVENTS) {
...
 
-
-    if (ArrayBuffer.isView(data)) {
-      // $FlowFixMe: no way to assert that 'data' is indeed an ArrayBufferView now
-      data = data.buffer;
-    }
-    if (data instanceof ArrayBuffer) {
-      data = base64.fromByteArray(new Uint8Array(data));
-      RCTWebSocketModule.sendBinary(data, this._socketId);
-      return;
```
Comment (ide) [inline Libraries/WebSocket/WebSocket.js]: Still need this return

Comment (philikon) [inline Libraries/WebSocket/WebSocket.js]: Duh, thanks

**Comment by ide** [comment]
@facebook-github-bot shipit

**Comment by philikon** [comment]
@mkonicek @foghina would you guys be able to take a look at why the tests are failing? I suspect it's due to some discrepancy between FB and OSS networking code.

**Comment by mkonicek** [comment]
Looks like the `com.facebook.react.CatalystNetworkingTestCase` failed (all tests). And now I see we haven't yet open sourced that test :/

Here's the test, it fails one line 63: http://pastebin.com/xjVprYma

**Comment by mkonicek** [comment]
Let's try once more just in case :)

@facebook-github-bot shipit

**Comment by philikon** [comment]
> Here's the test, it fails on lines 63, 75, 90: http://pastebin.com/xjVprYma

Thanks, unfortunately that doesn't tell me much. XHR works nicely in the OSS UIExplorer app. I suspect FB's custom networking stack, or XHRRequest, or whatever else is in the mix, needs to be slightly adopted for my change.

Could you apply this diff locally and run an internal Android app, see if the XHR works?

**Comment by philikon** [comment]
Sorry, never mind, I know what the problem is. Updated diff coming up.

**Comment by philikon** [comment]
Ok, pushed a fixed version. Could you re-import please?

**Comment by mkonicek** [comment]
@facebook-github-bot shipit

**Comment by philikon** [comment]
Sorry, I messed up last night. Rebased the wrong branch, pushed without my fix. Should be fixed now!

**Comment by ide** [comment]
@facebook-github-bot shipit

**Comment by philikon** [comment]
Yay thanks everyone!

**Comment by arv** [comment]
Will this be part of 0.41? It is not part of rc1 (afaict)

**Comment by ide** [comment]
@arv It will be part of 0.42. Since it landed before Feb 1, it will go out with the RC this week (cut around 1st of the month) and the final release a month after that. Sometimes we cherry pick small changes, particularly important bug fixes, but we prefer having commits this large and non-critical go through the regular release cycle and a full month of RC.

**Comment by arv** [comment]
Thanks @ide. I guess we will try 0.42rc then.

**Pull request closed without merging**